File: app/src/main/java/fi/tuska/jalkametri/gui/IconPickerDialog.java
package fi.tuska.jalkametri.gui;

import java.util.List;

import android.app.Dialog;
import android.content.Context;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.GridView;
import fi.tuska.jalkametri.Common;
import fi.tuska.jalkametri.R;
import fi.tuska.jalkametri.data.IconName;
import fi.tuska.jalkametri.util.ObjectCallback;

public class IconPickerDialog extends Dialog {

    private IconAdapter<IconName> adapter;
    private ObjectCallback<IconName> callback;

    public IconPickerDialog(Context context, ObjectCallback<IconName> callback) {
        super(context);
        this.callback = callback;
        setContentView(R.layout.select_icon);
        GridView list = (GridView) findViewById(R.id.list);

        List<IconName> icons = DrinkIconUtils.getAsList();
        adapter = new IconAdapter<IconName>(getContext(), icons, Common.DEFAULT_ICON_RES);
        list.setAdapter(adapter);

        list.setOnItemClickListener(new OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> a, View v, int position, long arg3) {
                selectIcon(adapter.getItem(position));
            }
        });
    }

    public void selectIcon(IconName icon) {
        if (callback != null)
            callback.objectSelected(icon);

        dismiss();
    }

}

File: app/src/main/java/fi/tuska/jalkametri/gui/IconPickerDialog.kt
package fi.tuska.jalkametri.gui

import android.app.DialogFragment
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView.OnItemClickListener
import android.widget.GridView
import fi.tuska.jalkametri.Common
import fi.tuska.jalkametri.Common.DEFAULT_ICON_RES
import fi.tuska.jalkametri.R
import fi.tuska.jalkametri.data.IconName
import java.io.Serializable

class IconPickerDialog : DialogFragment() {

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        val v = inflater.inflate(R.layout.select_icon, container, false)

        val callback = readArgs(arguments)
        fun selectIcon(icon: IconName) {
            callback(icon)
            this.dismiss()
        }

        val adapter = IconAdapter(v.context, DrinkIconUtils.getAsList(), DEFAULT_ICON_RES)
        val list = v.findViewById(R.id.list) as GridView
        list.adapter = adapter
        list.onItemClickListener = OnItemClickListener { _, _, position, _ -> selectIcon(adapter.getItem(position)) }

        return v
    }

    companion object {
        fun createDialog(callback: (IconName) -> Unit): IconPickerDialog = IconPickerDialog().apply {
            arguments = writeArgs(Bundle(), callback)
        }

        fun writeArgs(bundle: Bundle, callback: (IconName) -> Unit) = bundle.apply {
            putSerializable("callback", callback as Serializable)
        }

        @Suppress("UNCHECKED_CAST")
        fun readArgs(bundle: Bundle): (IconName) -> Unit =
                bundle.getSerializable("callback") as (IconName) -> Unit
    }

}
